### ISSUE TITLE
GWT RPC serialization for Dynamic Host Page

### ISSUE THREAD

**Issue by freddyboucher** [seanchenxi/gwt-storage#8]
Hi @seanchenxi 

My question is not directly related to **gwt-storage** but **gwt-storage** could be a part of the solution so can you take a look of this post plz?

http://stackoverflow.com/questions/27720895/gwt-rpc-serialization-for-dynamic-host-page

> I implemented a Dynamic Host Page (http://www.gwtproject.org/articles/dynamic_host_page.html) in my GWT project to pass my User POJO object directly in it.
> 
> I did it with AutoBean (https://code.google.com/p/google-web-toolkit/wiki/AutoBean) so I had to declare and implement Interfaces
> 
> But I found this way a little bit too verbose.
> 
> Is there another way? I mean GWT RPC do serialization without Interfaces so can I use it?
> 
> Another point; I already use the gwt-storage lib (https://github.com/seanchenxi/gwt-storage) that uses the GWT RPC mechanisim to store Java POJO Object in the Browser so there is probably an easy way...
> 
> THX

**Comment by freddyboucher** [comment]
Really thx to @rdwallis and @seanchenxi 

I posted my implementation on [Stackoverflow](http://stackoverflow.com/questions/27720895/gwt-rpc-serialization-for-dynamic-host-page/27759229#27759229)

@rdwallis 
I lost days because of the **//OK** that you have to remove.
I didn't read enough carefully your sample [RPC.encodeResponseForSuccess(...).substring(4);](https://github.com/rdwallis/alice/blob/567f12004393b98eeac71f47a210829d57f876b8/src/main/java/com/wallissoftware/alice/server/dispatch/AbstractActionHandler.java#L49)

**Comment by rdwallis** [comment]
@seachenxi, you're correct encodeResponseForSuccess is dangerous and will only serialize values that inherit IsSerializable.  This actually explains a lot of trouble I've been having with Enums.

So a server side serializer that takes into account Serialization Policy and removes the //OK from the front of the response would be very helpful.

**Comment by rdwallis** [comment]
@seanchenxi 
---
Sorry spelt your username incorrectly above and I know you don't always get notifications.

**Comment by seanchenxi** [comment]
@rdwallis 
That's ok. ;) And thank you very much for the precision.
I start working on this today.

**Comment by freddyboucher** [comment]
@rdwallis 
> This actually explains a lot of trouble I've been having with Enums.

How did you solve it? I'm facing the same issue with another project.

**Comment by rdwallis** [comment]
Make your enums implement IsSerializable.

**Comment by freddyboucher** [comment]
@rdwallis 
thx

**Comment by seanchenxi** [comment]
Hi, @rdwallis, @freddyboucher

I've added the serialization policy file generation in compilation process. Its path <gwt-module-name>/StorageSerializationPolicy.gwt.rpc in the latest snapshot.
So you could get it by using **com.google.gwt.user.server.rpc.SerializationPolicyLoader** in any servlet. 
The server side serializer is done, but not committed yet. I'm now working on a helper class which will be shared between client side and server side to simplify the serialization/deserialization work. 
Something like: 
In client side: String string = StorageUtils.serialize(myObject); 
In server side: String string = StorageUtils.serialize(myObject, "gwt-module-name"); // the second parameter will be optional.

If you have any other ideas, please tell me.

**Comment by rdwallis** [comment]
Awesome, will give it a try.

**Comment by freddyboucher** [comment]
Yep awesome, can't wait to test it.

**Comment by freddyboucher** [comment]
@seanchenxi 
I tried but **StorageSerializationPolicy.gwt.rpc** isn't generated :(

**Comment by seanchenxi** [comment]
@freddyboucher 
Really sorry. I've pushed a wrong snapshot version.
It should be fixed now. You can take a look at https://github.com/seanchenxi/gwt-storage-test

**Comment by freddyboucher** [comment]
@seanchenxi 
So I tested it and your sample app works fine! This is a terrific feature!!
But I'd like to call [StorageUtils.loadSerializationPolicy](https://github.com/seanchenxi/gwt-storage-test/blob/master/src/main/java/com/seanchenxi/gwt/storage/test/server/GreetingServiceImpl.java#L23) before my servlet **init(ServletConfig config)** because when you use a **jsp**, the init of your servlet is doing way after the process of your jsp.

I tried to call **StorageUtils.loadSerializationPolicy** directly in my **jsp** but at this time, the **StorageSerializationPolicy.gwt.rpc** isn't generated so it doesn't work.

When exactly the **StorageSerializationPolicy.gwt.rpc** is generated?

I've done a workaround right now to get it work: my **jsp** tests if my Servlet has been initialized. If yes it serializes my **User** if not it does nothing and I have to get my **User** by RPC request later. So when I launch my app only the first time ever my app does the RPC request and after that, if I reload my app my **jsp** does its job ;)

**Comment by seanchenxi** [comment]
Hi, @freddyboucher 

In classic dev mode, the StorageSerializationPolicy.gwt.rpc will be generated after the first use of StorageSerializer in client side. But in super dev mode, the file will be generated before the execution of jsp/html (thanks to the pre-compile).

I see the problem in your case, you must execute at least one time your client code for the file generation works. But this will not be a problem for production mode and super dev mode.

The serialization policy file generation was executed by the gwt's generator process. I'll try to find if there has a way to execute the file generation at dev mode startup.
